backend/src/main/kotlin/com/tngtech/apicenter/backend/connector/rest/service/SpecificationDataService.kt

```diff
@@ -74,24 +77,52 @@ class SpecificationDataService @Autowired constructor(
             null
         }
 
-    fun extractId(json: String): String? {
-        val parsedJson = objectMapper.readTree(json)
-        val jsonNode = parsedJson.at("/info/x-api-id")
-        return if (jsonNode.isMissingNode) {
+    fun extractId(json: String): String? =
+        try {
+            val parsedJson = objectMapper.readTree(json)
+            val jsonNode = parsedJson.at("/info/x-api-id")
+            if (jsonNode.isMissingNode) {
+                null
+            } else {
+                jsonNode.asText()
+            }
+        } catch (exception: IOException) {
             null
+        }
+
+    fun makeSpecificationMetaData(fileContent: String,
+                                  idFromPath: String? = null,
+                                  metaData: VersionFileMetaData? = null
+    ): VersionMetaData {
+        val idFromUpload = extractId(fileContent)
+        val serviceId = getConsistentServiceId(idFromUpload, idFromPath)
+
+        return if (metaData != null) {
+            VersionMetaData(
+                    serviceId,
+                    metaData.title,
+                    metaData.version,
+                    metaData.description,
+                    ApiLanguage.GRAPHQL,
+                    metaData.endpointUrl
+            )
         } else {
-            jsonNode.asText()
+            VersionMetaData(
+                    serviceId,
+                    extractTitle(fileContent),
+                    extractVersion(fileContent),
+                    extractDescription(fileContent),
+                    ApiLanguage.OPENAPI,
+                    null
+            )
         }
+
     }
 
-    fun makeSpecificationMetaData(fileContent: String, serviceId: ServiceId, endpointUrl: String?): VersionMetaData {
-        return VersionMetaData(
-                serviceId,
-                extractTitle(fileContent),
-                extractVersion(fileContent),
-                extractDescription(fileContent),
-                ApiLanguage.OPENAPI,
-                endpointUrl
-        )
+    private fun getConsistentServiceId(idFromUpload: String?, idFromPath: String?): ServiceId {
+        if (idFromUpload != null && idFromPath != null && idFromUpload != idFromPath) {
```

Comment:
`Objects.equals` should be able to do something very similar

Reply:
When both are null, the expression should return false (no exception thrown), whereas `Object.equals` would return true.